tests/php/Http/Controllers/Api/VolumeControllerTest.php

```diff
@@ -198,17 +235,420 @@ public function testUpdateValidation()
+        $id2 = $project->id;
+
+        $this->volume->projects()->attach($id2);
+
+        $this->doTestApiRoute('POST', "/api/v1/volumes/{$id}/project/{$id2}");
+
+        $this->beAdmin();
+
+        $img1 = ImageTest::create([
+            'lng' => 1.5,
+            'lat' => 5.3,
+            'filename' => 'a.jpg',
+            'volume_id' => $this->volume->id,
+        ]);
+        ImageLabelTest::create(['image_id' => $img1->id]);
+        ImageLabelTest::create(['image_id' => $img1->id]);
+        ImageLabelTest::create(['image_id' => $img1->id]);
+
+        $a = ImageAnnotationTest::create([
+            'image_id' => $img1->id,
+            'created_at' => '2010-01-01',
+        ]);
+        ImageAnnotationLabelTest::create([
+            'annotation_id' => $a->id,
+            'user_id' => $this->admin()->id,
+        ]);
+        $a2 = ImageAnnotationTest::create([
+            'image_id' => $img1->id,
+            'created_at' => '2015-03-03',
+        ]);
+        ImageAnnotationLabelTest::create([
+            'annotation_id' => $a2->id,
+            'user_id' => $this->admin()->id,
+        ]);
+
+        $img2 = ImageTest::create([
+            'lng' => 9.5,
+            'lat' => 9.0,
+            'filename' => 'b.jpg',
+            'volume_id' => $this->volume->id,
+        ]);
+        ImageLabelTest::create(['image_id' => $img2->id]);
+
+        $img3 = ImageTest::create([
+            'filename' => 'c.jpg',
+            'volume_id' => $this->volume->id,
+        ]);
+        $a3 = ImageAnnotationTest::create([
+            'image_id' => $img3->id,
+            'created_at' => '2000-09-10',
+        ]);
+        ImageLabelTest::create(['image_id' => $img2->id]);
+        ImageLabelTest::create(['image_id' => $img2->id]);
+
+        $a3 = ImageAnnotationTest::create([
+            'image_id' => $img3->id,
+            'created_at' => '2000-09-10',
+        ]);
+        ImageAnnotationLabelTest::create([
+            'annotation_id' => $a3->id,
+            'user_id' => $this->admin()->id,
+        ]);
+
+        $session = AnnotationSessionTest::create([
+            'volume_id' => $id,
+            'starts_at' => Carbon::today(),
+            'ends_at' => Carbon::tomorrow(),
+            'hide_own_annotations' => true,
+            'hide_other_users_annotations' => false,
+        ]);
+        $u1 = UserTest::create();
+        $u2 = UserTest::create();
+        $u3 = UserTest::create();
+        $session->users()->attach($this->admin()->id);
+        $session->users()->attach($u1);
+        $session->users()->attach($u2);
+        $session->users()->attach($u3);
+
+
+        $project->addUserId($this->admin()->id, Role::adminId());
+        Cache::flush();
+
+        $response = $this->post("/api/v1/volumes/{$id}/project/{$id2}");
+        $response->assertStatus(302);
+
+
+        $copy = $response->getSession()->get('copy');
+
+        $this->assertTrue($copy->images()->exists() == $this->volume->images()->exists());
```

Comment:
You use of `assertTrue()` can be problematic here in the whole function. If the argument evaluates to `false` you don't know which of the two compared values is not as expected. Also, the use of `==` instead of `===` could lead to unexpeced results. Only use `assertTrue()` if you actually want to test a scalar boolean. Use `assertEquals()` if you want to compare two values (even if they are boolean). The first argument of `assertEquals()` should be the expected value and the second argument should be the value to check.

Another issue I see here is that your assertion would pass even if `$this->volume` would be empty. But here you *know* that is shouldn't be empty. So in this particular case I would do:
```suggestion
        $this->assertTrue($copy->images()->exists());
```

Even better would be to actually check the attributes of the image while you are at it. This comes from my long comment above (the existence check is now done with `assertNotNull()`):
```php
$imageCopy = $volumeCopy->images()->first();
$this->assertNotNull($imageCopy);
$ignore = ['id', 'volume_id', 'uuid'];
$this->assertEquals(
    $image->makeHidden($ignore)->toArray(),
    $imageCopy->makeHidden($ignore)->toArray()
);
$this->assertNotEquals($image->uuid, $imageCopy->uuid);
```